### ISSUE TITLE
a combi logger method failed

### ISSUE THREAD

**Issue by BostonLobster** [MIC-DKFZ/trixi#20]
Hi, thanks for your excellent experiment framework! 

When I begin to use this framework, everything works fine. But accidently my $HOME directory runs out of disk space (which cause one of my jupyter kernel running trixi experiment failed). After clean and free the disk space, I run a trixi experiment from command line but get the following error when I want to add a numerical result. 

> a combi logger method failed:  <bound method ExperimentLogger.show_value of <trixi.logger.experiment.pytorchexperimentlogger.PytorchExperimentLogger object at 0x7f29552adba8>>

The same experiment worked fine yesterday, but when I re-run it, I get this weird result.

**Comment by BostonLobster** [comment]
After run `pip install --upgrade trixi`, I saw some packages are upgraded, but this error still exists. 

My code is originated from the official `pytorch_experiment.ipynb` example.

By adding `plot_results=False` in `self.add_result()`, this error is suppressed. I guess it is the problem of `self.plot_logger` in `experimentlogger.py`. 
Could anyone give some suggestions to solve this problem?

**Comment by jenspetersen** [comment]
Can you try the version on our develop branch? It's been rewritten recently.

**Comment by BostonLobster** [comment]
@jenspetersen Hi, I just tried the version on develop branch, and the previous problem is solved. But I encountered another problem this time... As shown below, when I use `show_text` to log training loss in terminal, an extra line with similar content is printed out in terminal too...

> INFO:default-qmEVYXDHoc:log: Train Epoch: 0 [340/5069 (7%)]     Landmark Loss: 27.8885, TV Loss: 0.1036
log: Train Epoch: 0 [360/5069 (7%)]     Landmark Loss: 66.0109, TV Loss: 0.0865
INFO:default-qmEVYXDHoc:log: Train Epoch: 0 [360/5069 (7%)]     Landmark Loss: 66.0109, TV Loss: 0.0865
log: Train Epoch: 0 [380/5069 (7%)]     Landmark Loss: 18.3065, TV Loss: 0.0983

How can I solve this? It is too annoying...

**Comment by dzimmerer** [comment]
Hi, this is a common issue. I think in the current pip visdom version this is a hiccup with the logging framework. Should be fixed and working if you install visdom from the current master.
Hope this can help you.

**Comment by jenspetersen** [comment]
To better understand the issue, here's an explanation :) This originates in the TextFileLogger that our (Pytorch)ExperimentLogger uses to log text to the terminal and to file. The TextFileLogger has an instance of a logging.Logger (that's the standard logging framework in Python), but with a random unique name, so we don't accidentally use a logger that one of the other libraries has already created. That logger now has a FileHandler (that writes to file) and a StreamHandler (that writes to stdout, i.e. your terminal). By default, a logging.Logger propagates messages to its parent logger, which is the root logger in our case (all logging.Logger instances are children or children of children etc. of the root logger). If some library we use (hint: it's Visdom) now does some funky business with the root logger, e.g. attach another StreamHandler like ours, you get those duplicate messages. I don't actually know what they do, but if you want to see it for yourself, open an IPython console or something and do this:

    exp = PytorchExperiment(base_dir="SOMETHING")
    exp.clog.show_text("test")

which should work fine. But now:

    exp = PytorchExperiment(base_dir="SOMETHING", loggers={"my_visdom_logger": "visdom"})
    exp.clog.show_text("test")

should give you that duplicate message.
As a quick fix, do this after instantiating your experiment:

    exp.elog.text_logger.loggers["default"].propagate = False

@dzimmerer we could make this the default, because I think most people use Visdom, but on the other hand I don't like the idea of fixing other people's mistakes in our own code :D

**Comment by BostonLobster** [comment]
@jenspetersen Many thanks for your detailed explaination! Now I can fix the error by following your suggestion.
If as you said, why don't you set `propagate` to `False` when `loggers` is not `None`?

**Comment by jenspetersen** [comment]
Because someone might want to do something else with the output logged to the trixi loggers in their environment. But I agree, that probably doesn't happen very often, and we might decide to do what you suggested...

**Comment by Larry-u** [comment]
Finally I find out it is not a mistake from yours, but a fault originated from matplotlib and the X11 forwarding.

#### The detailed explaination is as follow:
I run an experiment in terminal with X11 forwarding, and it runs well initially. But it suddenly fails with the following error: 

`XIO:  fatal IO error 25 (Inappropriate ioctl for device) on X server "localhost:10.0"
      after 49961 requests (49961 known processed) with 1921 events remaining`

Then I re-run this experiment, and it throws the `combi logger method failed` error. I further try to run `xclock` and `eog`, which both prompt 

`Error: Can't open display: localhost:10.0`

Then I type `xauth list`, the last line is 

`local/unix:11  MIT-MAGIC-COOKIE-1  ff95b396c585e7744a17fdae023cfea9`, note the `11` in this line. 

But  the `$DISPLAY` variable is still `localhost:10.0`, therefore the matplotlib will fail since it cannot open `localhost:10.0`.

Problem can be solved if I change `$DISPLAY` variable to `localhost:11.0`, which in line with the last line of `xauth list`. Then everything works fine.

Further solution may be changing the backend of matplotlib to non-gui one, because I really don't know when and how the authorization information and the `$DISPLAY` variable are mismatched, even if during the training.

**Comment by jenspetersen** [comment]
Seems unrelated? You can open a different issue and post this solution yourself, so it's less confusing for people arriving from Google with a similar error :)

**Comment by Larry-u** [comment]
@jenspetersen I am also trapped by this `combi logger method failed` error for a few days, and I finally find out the reason. I think it could be a reference for others:smile:

**Comment by jenspetersen** [comment]
@dzimmerer we could dump a trace instead of just printing the fact that a method failed :)
